docs: add the document for RWX volume support (#662)
harvester/docs: docs/rancher/csi-driver.md

1. You can now create a PVC based on this new **StorageClass**, which utilizes the **Host StorageClass** to provision volumes on the bare-metal Harvester cluster.

## RWX volumes support

### Prerequisites

> :::note
> 
> To use this feature, you need to ensure the host harvester cluster version is **v1.4.0** or later.
> 
> :::
> 
> Before creating RWX volumes on the guest cluster, you need to create a corresponding RWX StorageClass on the host cluster.
> 
> 1. Create a RWX StorageClass on the host

Reviewer: ```suggestion
- Harvester v1.4 or later is installed on the host cluster.

- You have created an RWX StorageClass on the host Harvester cluster.

  On the **Storage Class: Create** screen, click **Edit as YAML** and specify the following:
```

Reviewer: The Prerequisites section usually consists of a plain list. I am replacing both the note and the procedure with an unordered list.

  ```
  kind: StorageClass
  apiVersion: storage.k8s.io/v1
  metadata:
    name: longhorn-rwx
  provisioner: driver.longhorn.io
  allowVolumeExpansion: true
  reclaimPolicy: Delete
  volumeBindingMode: Immediate
  parameters:
    numberOfReplicas: "3"
    staleReplicaTimeout: "2880"
    fromBackup: ""
    fsType: "ext4"
    nfsOptions: "vers=4.2,noresvport,softerr,timeo=600,retrans=5"
  ```

  ![](/img/v1.4/rancher/create-rwx-sc-host-cluster-01.png)

  ![](/img/v1.4/rancher/create-rwx-sc-host-cluster-02.png)

  ![](/img/v1.4/rancher/create-rwx-sc-host-cluster-03.png)

> 2. Ensure the RBAC settings are up-to-date
> 
>   To support RWX feature, the harvester CSI driver needs to have the new RBAC settings. You can follow the steps below to check the RBAC settings.

Reviewer: ```suggestion
- The RBAC settings are up-to-date.

  The Harvester CSI driver requires must have the new RBAC settings to support RWX volumes. To check the RBAC settings, run the command `kubectl get clusterrole harvesterhci.io:csi-driver -o yaml`.
```

Reviewer: Which RBAC settings are you referring to? The Harvester doc does not contain any topics about RBAC settings. Please provide links to related information.

Author: The RBAC is the `Role-Based Access Control`. That is permission control to limit the resource operations (like `GET`, `LIST`, `CREATE` , `DELETE` and `UPDATE`).
The CSI driver needs some new permission to check/update the specific resource.

I will add some explanation here and provide the link.

Thanks!

Author: Updated. Please help to check, thanks!

    ```
    # kubectl get clusterrole harvesterhci.io:csi-driver -o yaml
    apiVersion: rbac.authorization.k8s.io/v1
    kind: ClusterRole
    metadata:
    ...
      name: harvesterhci.io:csi-driver
    ...
    rules:
    - apiGroups:
      - storage.k8s.io
      resources:
      - storageclasses
      verbs:
      - get
      - list
      - watch
    - apiGroups:
      - harvesterhci.io
      resources:
      - networkfilesystems
      - networkfilesystems/status
      verbs:
      - '*'
    - apiGroups:
      - longhorn.io
      resources:
      - volumes
      - volumes/status
      verbs:
      - get
      - list
    ```

> 3. Ensure the networkfs-manager pod(s) is running

Reviewer: ```suggestion
- The networkfs-manager pods are running.

  To check the status of the networkfs-manager pods, run the command `kubectl get pods -n harvester-system | grep networkfs-manager`.

  Example:
```

  ```
  # kubectl get pods -n harvester-system | grep networkfs-manager
  harvester-networkfs-manager-2pxhm                       1/1     Running     4 (34m ago)    3h41m
  harvester-networkfs-manager-8tst2                       1/1     Running     4 (37m ago)    3h41m
  harvester-networkfs-manager-xvkgp                       1/1     Running     4 (37m ago)    3h41m
  ```

> 4. Ensure the Harvester Version is at least v0.1.20
> 
>    If you use the older version of the Harvester CSI driver, upgrade it to v0.1.20 or later.

Reviewer: ```suggestion
- The Harvester CSI driver version is v0.1.20 or later.
```

  ![](/img/v1.4/rancher/harvester-csi-driver-version.png)

> ### Deploying and Usage
> 
> 1. Create a new StorageClass on the downstream cluster to associate with the RWX StorageClass on the host cluster

Reviewer: ```suggestion
### Usage

1. Create a new StorageClass on the guest cluster.

  On the **StorageClass: Create** screen, add a **Host Storage Class** parameter and specify the RWX StorageClass that you created on the host Harvester cluster.  
```

  
  ![](/img/v1.4/rancher/new-sc-associated-with-rwx.png)

> 2. Create RWX PVC with the new StorageClass

Reviewer: ```suggestion
1. Create an RWX PersistentVolumeClaim (PVC).

  On the **PersistentVolumeClaim: Create** screen, configure the following settings:
  
  - **Volume Claim** tab: Specify the new StorageClass.
  - **Customize** tab: Select **Many Nodes Read-Write**.
```

  ![](/img/v1.4/rancher/create-rwx-pvc-01.png)

  ![](/img/v1.4/rancher/create-rwx-pvc-02.png)

> 3. Verify the RWX PVC is created successfully

Reviewer: ```suggestion
1. Verify that the RWX PVC was created successfully.
```

Reviewer: How do users know that PVC creation was successful? Should they look for a specific state or status?

Author: The user usually checks with the GUI display on a downstream cluster, as shown in the following figure.

Reviewer: You didn't answer my question so I'll rephrase it. Which field on the GUI indicates that PVC creation was successful? Is it the state/status **Bound**?

Author: Yeah, sorry.

Check the status is **Bound** on the GUI.

Reviewer: Thanks!

  ![](/img/v1.4/rancher/check-rwx-pvc.png)

> 4. Create two pods to use the RWX PVC

Reviewer: ```suggestion
1. Create two pods.

  On the **Pod: Create** screen, specify the RWX PVC.
```

  ![](/img/v1.4/rancher/create-pod-with-rwx-pvc-01.png)

  ![](/img/v1.4/rancher/create-pod-with-rwx-pvc-02.png)

  ![](/img/v1.4/rancher/create-pod-with-rwx-pvc-03.png)

> Now, you can follow the same steps to create RWX PVC on the guest cluster. And use it on different pods which require the RWX volume.

Reviewer: ```suggestion
:::note

You can follow the same steps to create an RWX PVC on the guest cluster and then use it on pods that require RWX volumes.

:::
```
